### PR TITLE
Dev spam twirling evaluate

### PR DESCRIPTION
**Pull request by KilianPoirier**
## Description

- I fixed a known issue that the SPAM twirling feature has not implemented `expectation_with_uncertainty()`, calling `evaluate_circuit()` would not call the wrapper's redefinition of `expectation()` with the previous version of this code.
- The function `evaluate_circuit()` now functions correctly, further implementation of the uncertainty calculation with SPAM twirling is still possible.

## Checklist

[//]: <> (- [ ] My code follows the style guidelines of this project)

- [x] I have performed a self-review of my code.
- [x] I have commented my code and used numpy-style docstrings
- [ ] I have made corresponding updates to the documentation.
- [x] My changes generate no new warnings
- [ ] I have added/updated tests to make sure bugfix/feature works.
- [x] New and existing unit tests pass locally with my changes.

[/]: <> (- [ ] Any dependent changes have been merged and published in downstream modules)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py [hunk 1]
```diff
@@ -450,6 +450,21 @@ def evaluate_circuit(
...
+        elif isinstance(self.backend, SPAMTwirlingWrapper):
+            cost = self.backend.expectation(params_obj)
+            measurement_results = (
+                self.backend.measurement_outcomes
+                if isinstance(self.backend.measurement_outcomes, dict)
+                else self.backend.measurement_outcomes.tolist()
+            )
+            output_dict.update(
+                {
+                    "cost": cost,
```
Comment (vishal-ph) [inline src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py:464]: let's change the key `cost` to `expectation` in the `output_dict` for consistency with the docstring of the function

File: src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py [hunk 2]
```diff
@@ -450,6 +450,21 @@ def evaluate_circuit(
         # if the backend is the analytical simulator, we just return the expectation value of the cost Hamiltonian
         if isinstance(self.backend, QAOABackendAnalyticalSimulator):
             output_dict.update({"cost": self.backend.expectation(params_obj)[0]})
+        # if the workflow implements SPAM Twirling, 
+        # we just return the expectation value of the cost Hamiltonian and measurement outcomes
+        elif isinstance(self.backend, SPAMTwirlingWrapper):
+            cost = self.backend.expectation(params_obj)
+            measurement_results = (
+                self.backend.measurement_outcomes
```
Comment (vishal-ph) [inline src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py:458]: also,` measurement_results = ...` is repeated, so we can put it outside the if/else block

File: src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py (hunk 1 above)
Comment (KilianPoirier) [reply src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py:464]: Wouldn't it make more sense to change the docstring t `cost` instead?
I checked the implications of this change and the docstring is the only bit that mentions "uncertainty" as a field, or do you think the name `expectation` would be much clearer

File: src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py (hunk 2 above)
Comment (vishal-ph) [reply src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py:459]: @KilianPoirier, any updates on this?

File: src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py (hunk 1 above)
Comment (vishal-ph) [reply src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py:465]: Either is fine! You can make the choice, the main point is just consistency

File: src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py (hunk 2 above)
Comment (KilianPoirier) [reply src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_workflow.py:459]: `measurement_result` has to be calculated after calling `expectation()` or `expectation_with_uncertainty()` which means that combining the two would result in either an additional `if`, `else` statement or an additional dictionary update. Here we can clearly distinguish that for different backends we calculate different quantities and the result dictionary is only updated once.

**Review by vishal-ph — APPROVED** [review]

**Pull request merged**